### ISSUE TITLE
Initialization can be surprisingly important, and under-rated

### ISSUE THREAD

**Issue by hughperkins** [pytorch/pytorch#10243]
If you have a question or would like help and support, please ask at our
[forums](https://discuss.pytorch.org/).

If you are submitting a feature request, please preface the title with [feature request].
If you are submitting a bug report, please fill in the following details.

## Issue description

Was trying to reproduce a paper. Flatlined at 50% eval accuracy. On a two-class problem (binary cross-entropy).  Spent days pouring over their Tensorflow code, trying to figure out why their code worked, and mine didnt.

Eventually, turned to weight initialization. Used their initialization formula in my code. Bingo! 70% accuracy, same as theirs.

I think it might be good to make it easy/straight-forward to give a high quality initialization to LInear's and Embeddings.  The code for the paper I was reproducing draws weights uniformly from `[-r, r]`, where `r` is given by the following formula:

    r = factor * sqrt(3) / sqrt(num_inputs)

Where factor is:
- 1 for word embeddings
- 1.43 with a ReLU activation

And num_inputs is, for example for an `nn.Embedding(V, E)`, `num_inputs` would be `V`. For an `nn.Linear(V, E)`, `num_inputs` would also be `V`.

This formula is described also in Tensorflow, https://www.tensorflow.org/api_docs/python/tf/uniform_unit_scaling_initializer

A blog-post on the important of initialization, https://plus.google.com/+SoumithChintala/posts/RZfdrRQWL6u

I guess that it is unlikely that a one-size-fits-all approach will work for initialization (it's one more hyper-parameter to tune....), but might be good to at least make it visible/easy to switch between them? (Many people faced with my issue might have never tracked it down, and might have then thrown up their hands and said, "Well, some bug in pytorch; I'll just switch to Tensorflow...".)

Update (before posting):
- looks like there are init functions, at https://pytorch.org/docs/stable/nn.html#torch-nn-init
- however might be worth somehow making it more obvious/easier to use them in the nn.Linear and nn.Embedding api somehow? (currently, you have to read the source-code to find out default initialization; and you also have to kind of 'hack around' to pull the `.weight.data` out of the insides, to update them).

## Code example

Please try to provide a minimal example to repro the bug.
Error messages and stack traces are also helpful.

## System Info
Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

- PyTorch or Caffe2:
- How you installed PyTorch (conda, pip, source):
- Build command you used (if compiling from source):
- OS:
- PyTorch version:
- Python version:
- CUDA/cuDNN version:
- GPU models and configuration:
- GCC version (if compiling from source):
- CMake version:
- Versions of any other relevant libraries:

**Comment by SsnL** [comment]
Just want to point out that the initialization schemes are documented by @vishwakftw at https://github.com/pytorch/pytorch/pull/9038 . Thanks @vishwakftw !

**Comment by ezyang** [comment]
See also #18182

**Comment by mruberry** [comment]
Closing this due to age and because PyTorch's initialization schemes are now documented. If you have a more specific/actionable feature request or suggestion please let us know!

We agree: initialization is surprisingly important!
